k8s/zeppelin-server.yaml

```diff
@@ -0,0 +1,186 @@
+    daemon off;
+    worker_processes auto;
+    events {
+      worker_connections 1024;
+    }
+    http {
+      map $http_upgrade $connection_upgrade {
+        default upgrade;
+        '' close;
+      }
+
+      # first server block will be default. Proxy zeppelin server.
+      server {
+        listen 80;
+        location / {
+          proxy_pass http://localhost:8080;
+          proxy_set_header Host $host;
+          proxy_http_version 1.1;
+          proxy_set_header Upgrade $http_upgrade;
+          proxy_set_header Connection $connection_upgrade;
+          proxy_redirect http://localhost $scheme://SERVICE_DOMAIN;
+        }
+      }
+
+      # match request domain [port]-[service].[serviceDomain]
+      # proxy extra service such as spark-ui
+      server {
+        listen 80;
+        server_name "~(?<svc_port>[0-9]+)-(?<svc_name>[^.]*)\.(.*)";
+        location / {
+          resolver 127.0.0.1:53 ipv6=off;
+          proxy_pass http://$svc_name.NAMESPACE.svc.cluster.local:$svc_port;
+          proxy_set_header Host $host;
+          proxy_http_version 1.1;
+          proxy_set_header Upgrade $http_upgrade;
+          proxy_set_header Connection $connection_upgrade;
+          proxy_redirect http://localhost $scheme://SERVICE_DOMAIN;
+
+          # redirect rule for spark ui. 302 redirect response misses port number of service domain
+          proxy_redirect ~(http:[/]+[0-9]+[-][^-]+[-][^.]+)[^/]+(\/jobs.*) $1.SERVICE_DOMAIN$2;
+        }
+      }
+    }
+---
+kind: Pod
+apiVersion: v1
+metadata:
+  name: zeppelin-server
+  labels:
+    app: zeppelin-server
+spec:
+  automountServiceAccountToken: true
+  containers:
+  - name: zeppelin-server
+    image: apache/zeppelin:0.9.0-SNAPSHOT
+    command: ["sh", "-c", "$(ZEPPELIN_HOME)/bin/zeppelin.sh"]
+    lifecycle:
+      preStop:
+        exec:
+          # SIGTERM triggers a quick exit; gracefully terminate instead
+          command: ["sh", "-c", "ps -ef | grep org.apache.zeppelin.server.ZeppelinServer | awk '{print $2}' | xargs kill"]
+    env:
+    - name: ZEPPELIN_K8S_CONTAINER_IMAGE
+      value: apache/zeppelin:0.9.0-SNAPSHOT
+    - name: ZEPPELIN_HOME
+      value: /zeppelin
+    - name: ZEPPELIN_SERVER_RPC_PORTRANGE
+      value: 12320:12320
+    - name: POD_UID
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: metadata.uid
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: metadata.name
+    - name: ZEPPELIN_K8S_SPARK_CONTAINER_IMAGE
+      valueFrom:
+        configMapKeyRef:
+          name: zeppelin-server-conf
+          key: sparkContainerImage
+    - name: SERVICE_DOMAIN
+      valueFrom:
+        configMapKeyRef:
+          name: zeppelin-server-conf
+          key: serviceDomain
+    - name: MASTER   # default value of master property for spark interpreter.
+      value: k8s://https://kubernetes.default.svc
+  - name: zeppelin-server-gateway
+    image: nginx:1.14.0
```

Comment:
Is it better to create a svc for spark and use native Kubernetes redirection instead of ngnix as gateway?

Reply:
What i'm trying to do here, make Spark-ui accessible when user have access to Zeppelin ui, without any extra configuration/command to run.

I'm not familiar with native Kubernetes redirection. Could you explain how it can be possible with native Kubernetes redirection?

Reply:
Will this work for all modes (global, scoped, per-note, per-user)? I will try to check.
Implementing any security with embedded gateway might be a concern.

After some exploration, the change required will not be as simple as what has been provided. The approach could be as follows:  (We have used it for other applications)
- For the Spark-Driver UI, create a service with type=ClusterIP, port=randomPort, targetPort=driverUI port (4040 default). 
- Create an Ingress resource with path, service and port. Some annotations might be ingress controller specific sometimes. OR
- With this the spark-driver UI is accessible directly at the K8S edge-node/ 'path'.

I guess this would require some customization to be done by the users anyway.
If I get a chance to explore other ways, I will get back.